### ISSUE TITLE
[Feature request] Disable / control unmounting of already mounted rows

### ISSUE THREAD

**Issue by olee** [bvaughn/react-window#234]
I would like to suggest an option to control the unmounting of already mounted rows.
This could give a performance boost to lists with a moderate amount of large rows but which are very expensive to render.
If properly handled (rows should use memo of course) it would only render rows the first time they are visible and then keep them mounted to be able to scroll back to them again without needing to render them again.

What do you think of this suggestion?
Might there be a way to do this with current react-window (or maybe react-virtualized?) right now?

**Comment by bvaughn** [comment]
I don't think that sounds like something I want to support _directly_. It sounds like a reasonable use case, I just don't think I want to take on the extra complexity.

You could perhaps implement this behavior by supplying your own `innerElementType` that's stateful and maintains a list of rows it's created by inspecting its props:
https://github.com/bvaughn/react-window/blob/f2b2106cc828e9828cf2940411e5eab67f1d7aa6/src/createListComponent.js#L324-L331

But this would require some hacky duplication of this code:
https://github.com/bvaughn/react-window/blob/f2b2106cc828e9828cf2940411e5eab67f1d7aa6/src/createListComponent.js#L285-L298

**Comment by Bnaya** [comment]
Out of curiosity:
What's the painful part in your use-case? the DOM creation or react rendering?
I think another approach is implementing DOM reuse, but it will require adding intermediate level of virtual keys.

**Comment by olee** [comment]
Keeping react components mounted is pretty much the same as DOM reuse.
The initial rendering of new react nodes (ie. nodes that were added or have a new key) is the most expensive part of react rendering - subsequent updates are way faster.
This is why I would love to see a library supporting virtual scrolling, but at the same time keeping already rendered components mounted (because as long as they don't exceed a quite large amount, this should be faster).

PS: Oh yeah another option for "DOM reuse" might be a way to make the virtual scrolling **reuse the item keys**.
This should result in just their content being updated without making react disposing the old DOM elements and creating new ones.
However for this to work the items would have to be implemented in a way that ensures they do not totally rely on componentDidMount for some rendering logic.

**Comment by Bnaya** [comment]
@olee 
I'm exploring possibilities when exposing the up-to-date overscan information to the keyProvider and cell components (its a patch on react-window),
https://codesandbox.io/s/github/Bnaya/sticky-eurovision/tree/iteration
Take a look at `src/implementations/WithReactWindowDOMReuse/WithReactWindowDOMReuse.tsx`

Using custom keyProvider and modulo  i'm able to (almost) keep using same keys, and rotating them.

I think the idea behind it works, 
but right now react-window changing the overscan values in an unexpected way (for me) after the scroll ends. (losing one row and gaining it again right before next scroll)
I will keep exploring that direction

**Comment by Bnaya** [comment]
I've made another iteration, when i'm using component state to track recycling, looks very promising
In the picture we can see row 17 takes the components instances of row 2 
![Screen Shot 2019-06-06 at 17 13 02](https://user-images.githubusercontent.com/1304862/59040022-921ad680-887e-11e9-9c93-4b5b7f823538.png)

There is still an issue when you scroll up to the ends and the rendered rows number is changing.

**Comment by armavion** [comment]
I have the same problem here...
It would be great if there was a way which already-rendered items that fall out of the view, still be there and not unmount.

**Comment by bvaughn** [comment]
> It would be great if there was a way which already-rendered items that fall out of the view, still be there and not unmount.

Since this is something React intends to provide support for natively with a future Offscreen API, I don't think a library like react-window should try to implement this itself.

**Comment by olee** [comment]
I just thought that perhaps a way to achieve this could be outside of react-window by taking the virtualizedItems array from it and running it through some memoization code which keeps every previously rendered item by it's index and always just overwrites / adds the one passed into it.
Working on a PoC now...

**Comment by joseteIIo** [comment]
Agree with @codeaid . It's also a big pain point for a project I am a part of.

**Comment by erikiva** [comment]
Having the exact same problem as @codeaid with a List, is there any way around this?

**Comment by captainkovalsky** [comment]
If `overscanCount` equals the length of the items, will rows still be unmounted?

**Comment by CarlosVergikosk** [comment]
Same here, big pain.

**Comment by bvisserbb** [comment]
I haven't had a chance to try, but is there something that can be done with `rangeExtractor` -- to only expand the range of indexes returned and never shrink it?

**Comment by JustynaKK** [comment]
hi, did anybody found nice workaround? I have very expensive initial rendering and I am looking for the way to avoid unmounting

**Comment by CarlosVergikosk** [comment]
> hi, did anybody found nice workaround? I have very expensive initial rendering and I am looking for the way to avoid unmounting
 @JustynaKK use virtuoso.

**Comment by JustynaKK** [comment]
@CarlosVergikosk hi, I was checking virtuoso, but there is the same thing, everythig is unmounted on scroll
